src/MainWindow.vala

```diff
@@ -99,10 +104,42 @@ public class Mail.MainWindow : Gtk.ApplicationWindow {
             message_list_box.set_conversation (node);
         });
 
+        headerbar.size_allocate.connect (() => {
+            update_paned_start_grid_width ();
+            update_search_entry_width ();
+        });
+
+        paned_end.notify["position"].connect (update_search_entry_width);
+
+        paned_start.notify["position"].connect (() => {
+            update_paned_start_grid_width ();
+            update_search_entry_width ();
+        });
+
         Backend.Session.get_default ().start.begin ();
     }
 
     private void on_compose_message () {
         new ComposerWindow ().show_all ();
     }
+
+    private void update_search_entry_width () {
+        headerbar.search_entry.width_request = paned_end.position - paned_start.position + 1;
+    }
+    
+    private void update_paned_start_grid_width () {
+        int offset = 0;
+        headerbar.forall ((widget) => {
+            if (widget.get_style_context ().has_class ("left")) {
+                Gtk.Allocation alloc;
+                widget.get_allocation (out alloc);
+                offset = alloc.width;
```

Comment:
We should probably add a `return` at the end here to save some more of those CPU cycles. My bad.